firedrake/plot.py

```diff
@@ -1,373 +1,279 @@
     """
-    import matplotlib.pyplot as plt
-    try:
-        from ipywidgets import interact, IntSlider
-    except ImportError:
-        raise RuntimeError("Not in notebook")
+    return _plot_2d_field("tricontour", function, *args, **kwargs)
 
+
+def trisurf(function, *args, **kwargs):
+    r"""Create a 3D surface plot of a finite element field
+
+    :arg function: the function to plot
+    :return:
+    """
+    axes = kwargs.pop("axes", None)
     if axes is None:
-        axes = plt.subplot(111)
+        figure = plt.figure()
+        axes = figure.add_subplot(projection='3d')
 
-    def display_plot(index):
-        axes.clear()
-        return plot(functions[index], num_sample_points, axes=axes, **kwargs).figure
+    if len(function.ufl_shape) == 1:
+        mesh = function.ufl_domain()
+        element = function.ufl_element().sub_elements()[0]
+        Q = FunctionSpace(mesh, element)
+        function = interpolate(sqrt(inner(function, function)), Q)
 
-    interact(display_plot, index=IntSlider(min=0, max=len(functions)-1,
-                                           step=1, value=0),
-             continuous_update=False)
-    return axes
+    num_sample_points = kwargs.pop("num_sample_points", 10)
+    triangulation, vals = _two_dimension_triangle_func_val(function,
+                                                           num_sample_points)
 
+    _kwargs = {"antialiased": False, "edgecolor": "none", "shade": False,
+               "cmap": plt.rcParams["image.cmap"]}
+    _kwargs.update(kwargs)
 
-def piecewise_linear(points, axes=None, **kwargs):
-    """Plot a piece-wise linear plot for the given points, returns a
-    matplotlib axes.
+    return axes.plot_trisurf(triangulation, vals, *args, **_kwargs)
 
-    :arg points: Points to be plotted
-    :arg axes: Axes to be plotted on
-    :arg kwargs: Additional key word arguments passed to plot
+
+def tripcolor(function, *args, **kwargs):
+    r"""Create a pseudo-color plot of a finite element field
+
+    :arg function: the function to plot
+    :return: matplotlib Collection object
+    """
+    return _plot_2d_field("tripcolor", function, *args, **kwargs)
+
+
+def quiver(function, *args, **kwargs):
+    r"""Make a quiver plot of a vector field
+
+    :arg function: the vector field to plot
+    :return: matplotlib Quiver object
     """
-    try:
-        import matplotlib.pyplot as plt
-    except ImportError:
-        raise RuntimeError("Matplotlib not importable, is it installed?")
+    if function.ufl_shape != (2,):
+        raise ValueError("Quiver plots only defined for 2D vector fields!")
+
+    axes = kwargs.pop("axes", None)
+    if axes is None:
+        figure = plt.figure()
+        axes = figure.add_subplot()
+
+    coords = function.ufl_domain().coordinates.dat.data_ro
+    vals = np.asarray(function.at(coords, tolerance=1e-10))
+    C = np.linalg.norm(vals, axis=1)
+    return axes.quiver(*(coords.T), *(vals.T), C, *args, **kwargs)
```

Comment:
For the quiver plots, we just interpolate the function at the coordinates and don't do any sub-sampling. I wonder if we should do the same thing for the triangulations (especially now that paraview output supports high-order viz)?

WDYT?

Reply:
I'm not sure I understand the suggestion -- are you thinking we should use `function.at` on the mesh coordinates instead of subsampling for functions like tricontour, tricontourf, etc? Right now these functions call out to `_two_dimension_triangle_func_val` and then to `_calculate_values`. If it's ok with you I'd rather use this PR to change the API to match matplotlib, and then we can worry about simplifying the internal workings of the code later.

Specifically for quiver, subsampling is a bad idea because the spacing of the arrows can be so dense that the plot is impossible to parse visually.

Reply:
At least we should not be using `function.at`.  `at` discards all the mesh topology information and does geometric searches for elements. This is structurally slow. Instead one should use:
`vals = interpolate(function, coords.function_space())`

Reply:
I thought the use of `.at` there was a little fishy -- I fixed it as per your suggestion.